8315149: Add hsperf counters for CPU time of internal GC threads

jjoo172 opened the pull request:
<!--
Replace this text with a description of your pull request (also remove the surrounding HTML comment markers).
If in doubt, feel free to delete everything in this edit box first, the bot will restore the progress section as needed.
-->

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8315149](https://bugs.openjdk.org/browse/JDK-8315149): Add hsperf counters for CPU time of internal GC threads (**Enhancement** - P4)


### Reviewers
 * [Volker Simonis](https://openjdk.org/census#simonis) (@simonis - **Reviewer**) ⚠️ Review applies to [9fb36a9e](https://git.openjdk.org/jdk/pull/15082/files/9fb36a9e4ef5bffe86feebf18e704db38101692d)
 * [Man Cao](https://openjdk.org/census#manc) (@caoman - Committer)
 * [Stefan Johansson](https://openjdk.org/census#sjohanss) (@kstefanj - **Reviewer**)


### Contributors
 * Man Cao `<manc@openjdk.org>`
 * Stefan Johansson `<sjohanss@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15082/head:pull/15082` \
`$ git checkout pull/15082`

Update a local copy of the PR: \
`$ git checkout pull/15082` \
`$ git pull https://git.openjdk.org/jdk.git pull/15082/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15082`

View PR using the GUI difftool: \
`$ git pr show -t 15082`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15082.diff">https://git.openjdk.org/jdk/pull/15082.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15082#issuecomment-1698150598)

dholmes-ora commented on src/hotspot/share/gc/g1/g1ConcurrentMark.cpp:
```diff
@@ -453,6 +453,14 @@ G1ConcurrentMark::G1ConcurrentMark(G1CollectedHeap* g1h,
 
   reset_at_marking_complete();
+
+  if (UsePerfData) {
```
Should this also check `os::is_thread_cpu_time_supported()`?

dholmes-ora commented on src/hotspot/share/gc/g1/g1ConcurrentMark.hpp:
```diff
@@ -371,6 +374,7 @@ class G1ConcurrentMark : public CHeapObj<mtGC> {
   void verify_during_pause(G1HeapVerifier::G1VerifyType type,
                            VerifyLocation location);
 
+
```
Spurious blank line added.

dholmes-ora commented on src/hotspot/share/gc/g1/g1ConcurrentRefine.cpp:
```diff
@@ -112,6 +113,13 @@ jint G1ConcurrentRefineThreadControl::initialize(G1ConcurrentRefine* cr, uint ma
     }
   }
 
+  if (UsePerfData) {
```
Again should we only initialize if also `os::is_thread_cpu_time_supported()`?

dholmes-ora commented on src/hotspot/share/gc/g1/g1ConcurrentRefine.cpp:
```diff
@@ -152,6 +160,22 @@ void G1ConcurrentRefineThreadControl::stop() {
+void G1ConcurrentRefineThreadControl::update_threads_cpu_time() {
+  // The primary thread (_threads[0]) updates the counter for all worker
+  // threads, because:
+  // the primary thread is always waken up first from being blocked on a monitor
```
s/waken/woken/

dholmes-ora commented on src/hotspot/share/gc/g1/g1ConcurrentRefine.cpp:
```diff
@@ -152,6 +160,22 @@ void G1ConcurrentRefineThreadControl::stop() {
+  assert(Thread::current() == _threads[0],
+         "Must be called from G1ConcurrentRefineThreadControl::_threads[0] to "
+         "avoid races");
+  assert(UsePerfData, "Must be enabled");
```
`&& os::is_thread_cpu_time_supported()` ?

dholmes-ora reviewed (commented):
This doesn't seem unreasonable, but I do wonder what impact this has on the GC threads to collect this data?

I mentioned it a couple of times, but generally speaking should we only initialize these perf counters when `os::is_thread_cpu_time_supported()`? Or do they need to be readable regardless?

Thanks.

caoman commented:
> This doesn't seem unreasonable, but I do wonder what impact this has on the GC threads to collect this data?

We did some benchmarking when we first implemented these counters in our internal JDK 8. There was no observable performance difference. We could redo the benchmarking in JDK HEAD if people are concerned with accounting overhead.

dholmes-ora commented on src/hotspot/share/gc/shared/collectedHeap.cpp:
```diff
@@ -274,6 +274,10 @@ CollectedHeap::CollectedHeap() :
                 PerfDataManager::create_string_variable(SUN_GC, "lastCause",
                              80, GCCause::to_string(_gc_lastcause), CHECK);
+
+    _perf_parallel_gc_threads_cpu_time =
+                PerfDataManager::create_variable(NULL_NS, "par_gc_thread_time",
+                                                 PerfData::U_Ticks, CHECK);
```
Shouldn't this initialization be guarded like the others?

dholmes-ora commented on src/hotspot/share/runtime/vmThread.cpp:
```diff
@@ -136,6 +137,9 @@ void VMThread::create() {
     _perf_accumulated_vm_operation_time =
                  PerfDataManager::create_counter(SUN_THREADS, "vmOperationTime",
                                                  PerfData::U_Ticks, CHECK);
+    _perf_vm_thread_cpu_time =
+                 PerfDataManager::create_variable(NULL_NS, "vm_thread_time",
+                                                  PerfData::U_Ticks, CHECK);
```
Again should not this initialization be guarded like the others.

dholmes-ora commented on src/hotspot/share/runtime/vmThread.hpp:
```diff
@@ -123,6 +124,9 @@ class VMThread: public NamedThread {
   static PerfCounter* perf_accumulated_vm_operation_time() {
     return _perf_accumulated_vm_operation_time;
   }
+  static PerfVariable* perf_vm_thread_cpu_time() {
+    return _perf_vm_thread_cpu_time;
+  }
```
Why do you need an accessor for the cpu-time variable in the VMThread but none of the others?

dholmes-ora commented on src/hotspot/share/gc/shared/stringdedup/stringDedupThread.hpp:
```diff
@@ -40,6 +40,9 @@
   friend class VMStructs;
 
+  // Perf data for CPU time consumed by the string dedup thread.
+  static PerfVariable* _g1_concurrent_dedup_thread_cpu_time;
```
This variable seems unused. ??

caoman replied:
It's better to move this variable to stringDedupProcessor.hpp/cpp. Then it will be updated in `StringDedup::Processor::run()`.

Also, we should rename this variable so it is not specific to G1. Shenandoah also supports string dedup (https://bugs.openjdk.org/browse/JDK-8203641).

simonis commented on src/hotspot/share/gc/g1/g1ConcurrentMark.cpp:
```diff
@@ -453,6 +453,14 @@ G1ConcurrentMark::G1ConcurrentMark(G1CollectedHeap* g1h,
+    EXCEPTION_MARK;
+
+    _g1_concurrent_mark_threads_cpu_time =
+        PerfDataManager::create_variable(NULL_NS, "g1_conc_mark_thread_time",
```
See my general comment about name spaces. The name should be something like `*.cpu_time`.

caoman replied:
+1. We probably should avoid using `NULL_NS` as well. So perhaps something like:
```
PerfDataManager::create_variable(SUN_THREADS, "g1_conc_mark_cpu_time", ...
```
There is only one existing hsperf counter under `SUN_THREADS`: sun.threads.vmOperationTime.
I think it is appropriate to add all these thread CPU time counters under `SUN_THREADS`.

simonis commented on src/hotspot/share/gc/g1/g1ConcurrentRefine.cpp:
```diff
@@ -112,6 +113,13 @@ jint G1ConcurrentRefineThreadControl::initialize(G1ConcurrentRefine* cr, uint ma
+  if (UsePerfData && os::is_thread_cpu_time_supported()) {
+    EXCEPTION_MARK;
+    _g1_concurrent_refine_threads_cpu_time =
+        PerfDataManager::create_variable(NULL_NS, "g1_conc_refine_thread_time",
```
See my general comment about name spaces. The name should be something like `*.cpu_time`.

simonis commented on src/hotspot/share/gc/shared/collectedHeap.cpp:
```diff
@@ -274,6 +274,10 @@ CollectedHeap::CollectedHeap() :
                              80, GCCause::to_string(_gc_lastcause), CHECK);
+
+    _perf_parallel_gc_threads_cpu_time =
+                PerfDataManager::create_variable(NULL_NS, "par_gc_thread_time",
```
See my general comment about name spaces. The name should be something like `*.cpu_time`.

simonis commented on src/hotspot/share/gc/shared/collectedHeap.hpp:
```diff
@@ -141,6 +142,10 @@ class CollectedHeap : public CHeapObj<mtGC> {
 
+  // Perf counters for CPU time of parallel GC threads. Defined here in order to
+  // be reused for all collectors.
+  PerfVariable* _perf_parallel_gc_threads_cpu_time;
```
If this is intended to be reused for other GCs then rename to something more generic like `_perf_gc_threads_cpu_time`.

caoman replied:
The intention is for parallel GC worker threads. Not for all GC threads, and not to be confused with threads for `ParallelGC`.
Perhaps `_perf_parallel_worker_threads_cpu_time` is better?

simonis commented on src/hotspot/share/gc/shared/collectedHeap.hpp:
```diff
@@ -551,4 +556,30 @@ class GCCauseSetter : StackObj {
+// Class to compute the total CPU time for a set of threads, then update an
+// hsperfdata counter.
+
+class ThreadTotalCPUTimeClosure: public ThreadClosure {
```
This class is not really related to GC (and you are already using it to get the CPU time of the `VMThread`) so I wonder if we should put this in a more generic location (maybe `thread.hpp` or something similar). This will become even more relevant if we implement CPU time for JIT compiler threads because I don't think we want to have a dependency on `collectedHeap.hpp` there.

simonis commented on src/hotspot/share/runtime/vmThread.cpp:
```diff
@@ -136,6 +137,9 @@ void VMThread::create() {
                  PerfDataManager::create_counter(SUN_THREADS, "vmOperationTime",
                                                  PerfData::U_Ticks, CHECK);
+    _perf_vm_thread_cpu_time =
+                 PerfDataManager::create_variable(NULL_NS, "vm_thread_time",
```
Please move to the corresponding namespace and rename to something like `*.cpu_time`.

simonis reviewed (changes requested):
First of all thanks for this PR. I think it is useful and once it's done I think we're interested in implementing the corresponding counters for Shenandoah and also for the JIT compiler threads.

First some general comments:
- Instead of `PerfVariable` I'd use `PerfCounter` because they represent a "[*data value that can (should) be modified in a monotonic manner*](https://github.com/openjdk/jdk/blob/bd477810b176696e0fd043f5594663ebcf9884cf/src/hotspot/share/runtime/perfData.hpp#L419-L427)" rather than "*being modified in an unrestricted manner*".
- Please put the counters into the appropriate namespace (e.g. `sun.gc.collector.<n>.cpu_time`). This fits better with the existing counters (we currently don't have counters outside the `java.` or `sun.` namespaces) and makes it easier for follow up changes to implement the corresponding counters for additional GCs.
- Can you please aggregate all the different CPU time counters from the different GC phases into one GC CPU time counter (e.g. `sun.gc.cpu_time`). This would be the same for different GCs and would simplify the monitoring of GC overhead independently of the used GC algorithm.
- Please add a test for the new counters.

caoman commented on src/hotspot/share/gc/g1/g1ConcurrentRefine.cpp:
```diff
@@ -152,6 +160,22 @@ void G1ConcurrentRefineThreadControl::stop() {
+  // G1ConcurrentRefineThread's constructor);
+  // the primary thread is started last and stopped first, so it will not risk
+  // reading CPU time of a terminated worker thread.
+  assert(Thread::current() == _threads[0],
```
This assert could be changed to call `assert_current_thread_is_primary_refinement_thread()`.

caoman commented on src/hotspot/share/gc/g1/g1ConcurrentRefineThread.cpp:
```diff
@@ -79,6 +79,10 @@ void G1ConcurrentRefineThread::run_service() {
+
+    if (UsePerfData && os::is_thread_cpu_time_supported() && is_primary()) {
+      _cr->update_concurrent_refine_threads_cpu_time();
+    }
```
There are two classes for primary thread and secondary refinement thread: `G1PrimaryConcurrentRefineThread` and `G1SecondaryConcurrentRefineThread`. It is probably cleaner to move this part inside `G1PrimaryConcurrentRefineThread` and add a virtual method in `G1ConcurrentRefineThread`. We can get rid of the `is_primary()` check as well.

```
class G1ConcurrentRefineThread {
   virtual void possibly_update_threads_cpu_time() {};
}

void G1PrimaryConcurrentRefineThread::possibly_update_threads_cpu_time() {
  if (UsePerfData && os::is_thread_cpu_time_supported()) {
    _cr->update_concurrent_refine_threads_cpu_time();
  }
}
```

jjoo172 replied:
Sounds good, let me know if this change is how you envisioned it!

caoman commented:
Responding to some comments from @simonis:

> Please put the counters into the appropriate namespace (e.g. sun.gc.collector.<n>.cpu_time).

I suggested using the `SUN_THREADS` namespace instead, because:
- Counters like for VM thread and String Dedup thread do not belong to any GC collector namespace.
- It is hard to classify some counters into either `sun.gc.collector.0` or `sun.gc.collector.1`. E.g. `_perf_parallel_gc_threads_cpu_time` is used by both incremental and full collectors for G1. It is possible to put them under `sun.gc` directly though.

> Can you please aggregate all the different CPU time counters from the different GC phases into one GC CPU time counter (e.g. sun.gc.cpu_time).

Our experience for monitoring metrics is that it is best to provide individual metrics for each subcomponent. Monitoring tools can perform aggregation for these fine-grained metrics if they need to. If the JVM provide an aggregate metric, it is impossible to un-aggregate this metric. It is also hard to tell which subcomponent is the culprit if there's regression.

caoman reviewed (changes requested):
Have you enabled Github actions on your personal fork? I don't see the OpenJDK pre-submit build and tests being executed.

See https://wiki.openjdk.org/display/SKARA/Testing#Testing-Configuringworkflowstorun, https://bugs.openjdk.org/browse/SKARA-846

simonis reviewed (changes requested):
Thanks for your changes. Already much better now, but I still have some comments :)

caoman commented on src/hotspot/share/gc/g1/g1ConcurrentMark.cpp:
```diff
@@ -453,6 +453,14 @@ G1ConcurrentMark::G1ConcurrentMark(G1CollectedHeap* g1h,
+    EXCEPTION_MARK;
+
+    _g1_concurrent_mark_threads_cpu_time =
+        PerfDataManager::create_counter(SUN_THREADS, "g1_conc_mark_thread.cpu_time",
```
Thanks for the clarification. Yes, adding a new counter for aggregation sounds good.

jjoo172 commented:
/contributor add manc

simonis commented:
/label add serviceability

dholmes-ora commented on src/hotspot/share/gc/g1/g1CollectedHeap.cpp:
```diff
@@ -2414,6 +2416,24 @@ void G1CollectedHeap::verify_region_attr_remset_is_tracked() {
+  if (worker_threads != NULL) {
+    ThreadTotalCPUTimeClosure tttc(_perf_parallel_worker_threads_cpu_time, true);
+    // Currently parallel worker threads never terminate (JDK-8081682), so it is
+    // safe for VMThread to read their CPU times. If upstream fixes JDK-8087340
```
The reference to "upstream" is not applicable here - this is "upstream".

jjoo172 commented on test/jdk/sun/tools/jcmd/TestGcCounters.java:
```diff
@@ -0,0 +1,40 @@
+
+        output.shouldHaveExitValue(0);
+        output.shouldContain(SUN_THREADS + ".total_gc_cpu_time");
+        output.shouldContain(SUN_THREADS_GCCPU + ".g1_conc_mark");
```
Is the proper way to do this via adding the @requires annotation in the test header, like done in https://github.com/openjdk/jdk/blob/8f46abc938ffe338e25d5fdbdcfa0aaa12edfa58/test/hotspot/jtreg/gc/stress/TestStressG1Humongous.java#L30?

dholmes-ora replied:
Yes

dholmes-ora reviewed (commented):
Thanks for updates.

dholmes-ora commented on src/hotspot/share/gc/shared/collectedHeap.cpp:
```diff
@@ -157,12 +158,21 @@ bool CollectedHeap::contains_null(const oop* p) const {
 }
 
-void CollectedHeap::inc_total_cpu_time(jlong diff) {
+void CollectedHeap::inc_total_cpu_time(long diff) {
```
We don't use `long` in shared code as it has different size on different platforms.

caoman replied:
Using `long` is to avoid build failure on 32-bit ARM and x86. `jlong` is `long long` on 32-bit, and Atomic template does not support `long long` on 32-bit. Example failure: https://github.com/jjoo172/jdk/actions/runs/6229455243/job/16907994694.

Is there a better way to avoid these failures on 32-bit?

dholmes-ora replied:
`long` is 32-bit on Windows x64 as well which means you're reducing the utility of these timers there (else you could use 32-bit everywhere).

AFAICS it should be supported on x86-32 as we define `SUPPORTS_NATIVE_CX8` whilst for ARM it is restricted to ARMv7a and above. (Does anyone build ARMv6 still?) But that appears not to be handled by the atomic templates.

Not sure the best way to approach this one. If the templates correctly handled SUPPORTS_NATIVE_CX8 to define the 64-bit variants then the ideal solution would be to use a typedef that is 64-bit on supported platforms and 32-bit elsewhere.

shipilev replied:
Looks to me that x86_32 actually implements PlatformCmpxchg<8> (via linux_x86.S assembly), but not PlatformAdd<8>. So maybe the workaround would be to use Atomic::cmpxchg for this update, at least on _LP64 path? I see CollectedHeap::publish_total_cpu_time already does the CAS too.

simonis commented on src/hotspot/share/gc/shared/stringdedup/stringDedupThread.cpp:
```diff
@@ -28,6 +28,7 @@
 #include "gc/shared/stringdedup/stringDedupThread.hpp"
 #include "runtime/handles.hpp"
 #include "runtime/os.hpp"
+#include "runtime/perfData.hpp"
```
This include doesn't seem to be needed?

simonis reviewed (commented):
Thanks for your continued work on this PR. I think it looks pretty good now once you fix the `long`/`jlong` issue mentioned by @dholmes-ora. And the solution proposed by @shipilev should work equally well on arm32.

jjoo172 commented:
Resolved comments and sanity checks pass on all builds: https://github.com/jjoo172/jdk/actions/runs/6411637099

I believe this PR should be RFR once again.

caoman commented on src/hotspot/share/gc/shared/collectedHeap.cpp:
```diff
@@ -264,7 +295,7 @@ CollectedHeap::CollectedHeap() :
   NOT_PRODUCT(_promotion_failure_alot_gc_number = 0;)
 
-  if (UsePerfData) {
+  if (UsePerfData && os::is_thread_cpu_time_supported()) {
```
This condition should be a nested if inside `if (UsePerfData)`:

```
if (os::is_thread_cpu_time_supported()) {
  _total_cpu_time = ...;
  _perf_parallel_worker_threads_cpu_time = ...;
}
```
Otherwise `_perf_gc_cause` and `_perf_gc_lastcause` could be broken.

jjoo172 replied:
Ah yes, good catch!

caoman commented on src/hotspot/share/gc/shared/collectedHeap.hpp:
```diff
@@ -194,6 +208,7 @@ class CollectedHeap : public CHeapObj<mtGC> {
     Shenandoah
   };
 
+
```
Nit: unnecessary new line.

caoman commented on src/hotspot/share/gc/parallel/parallelScavengeHeap.cpp:
```diff
@@ -884,3 +885,16 @@ void ParallelScavengeHeap::pin_object(JavaThread* thread, oop obj) {
+  // is safe for VMThread to read their CPU times. If upstream changes this
+  // behavior, we should rethink if it is still safe.
+  gc_threads_do(&tttc);
+}
```
It should also call `CollectedHeap::publish_total_cpu_time()` here, right?

caoman commented on src/hotspot/share/runtime/vmThread.cpp:
```diff
@@ -130,12 +131,15 @@ void VMThread::create() {
   _terminate_lock = new Monitor(Mutex::nosafepoint, "VMThreadTerminate_lock");
 
-  if (UsePerfData) {
+  if (UsePerfData && os::is_thread_cpu_time_supported()) {
```
Similarly, check for `os::is_thread_cpu_time_supported()` should be a inner nested if, to avoid breaking `_perf_accumulated_vm_operation_time`.

caoman commented on test/jdk/sun/tools/jcmd/TestGcCounters.java:
```diff
@@ -0,0 +1,41 @@
+ */
+public class TestGcCounters {
+
+    private static final String[] VM_ARGS = new String[] { "-XX:+UsePerfData" };
```
The `VM_ARGS` is passed to the jcmd subprocess via `JcmdBase.jcmd()`. This is probably not what we want, we want to pass `-XX:+UsePerfData` to the main process that runs `TestGcCounters.main()`.

We probably need this on line 13:
```
@run main/othervm -XX:+UsePerfData TestGcCounters
```

jjoo172 commented on src/hotspot/share/runtime/perfData.hpp:
```diff
@@ -61,6 +61,7 @@ enum CounterNS {
   JAVA_THREADS,         // Threads System name spaces
   COM_THREADS,
   SUN_THREADS,
+  SUN_THREADS_GCCPU,    // Subsystem for Sun Threads GC CPU
```
Updated the counters for now to these names, but open to renaming again based on feedback from @simonis!

shipilev commented:
FYI, 64-bit atomics fallbacks for 32-bit arches are developing here: #16252.

simonis commented:
This looks good to me now. I only propose to wait until #16252 will be pushed (it already has two reviews) and than rebase on top of that change and use the generic 64-bit atomics instead of your handcrafted solution. That will remove the need for the clumsy `#ifdef _LP64` code.

Once that is done, I'll approve this PR.

albertnetymk commented:
Since it introduces new APIs in `CollectedHeap`...

/label add hotspot-gc

simonis commented:
@jjoo172, https://github.com/openjdk/jdk/pull/16252 has now been integrated. Can you please merge it in and use the 64-bit Atomics for incrementing the counters?

caoman commented on src/hotspot/share/gc/shared/stringdedup/stringDedupProcessor.cpp:
```diff
@@ -187,6 +196,10 @@ void StringDedup::Processor::run(JavaThread* thread) {
     _cur_stat.report_active_end();
     log_statistics();
+    if (UsePerfData && os::is_thread_cpu_time_supported()) {
+      ThreadTotalCPUTimeClosure tttc(_concurrent_dedup_thread_cpu_time, true);
```
I think it is better to not classify StringDedup thread as a GC thread, so remove the "true" parameter.
Although StringDedup requests are initiated during GC (search `_string_dedup_requests.add`), they are not part of the GC.

caoman reviewed (approved):
The change LGTM except a small suggestion.
Could some one from hotspot-gc-dev also take a look as @albertnetymk added the tag?

simonis reviewed (approved):
Thanks for your patience :)

Looks good to me.

albertnetymk commented on src/hotspot/share/gc/g1/g1ConcurrentMarkThread.cpp:
```diff
@@ -134,6 +134,8 @@ void G1ConcurrentMarkThread::run_service() {
 
     _vtime_accum = (os::elapsedVTime() - _vtime_start);
+
+    cm()->update_concurrent_mark_threads_cpu_time();
```
Is there some overlapping btw this and the existing `_vtime_accum`. If so, can they be consolidated somehow?

I believe the purpose of calling `update_concurrent_mark_threads_cpu_time` in multiple places is to get more up-to-date conc-cpu-time. Reading through the JBS ticket, I don't see the motivation for maintaining such a "fresh" value.

Finally, is CSR required for this feature?

simonis replied:
@albertnetymk, the hsperf counters are a non-public API and the new counters have been added to the non-standard `sun.threads.cpu_time` name space which is "[unstable and unsupported](https://github.com/openjdk/jdk/blob/9864951dceb0ddc4479ced04b6d5a2363f1e307d/src/hotspot/share/runtime/perfData.cpp#L56)" so I don't think an CSR is required.

caoman replied:
> Is there some overlapping btw this and the existing _vtime_accum. If so, can they be consolidated somehow?

We don't really like `_vtime_accum` for monitoring:
- `os::elapsedVTime()` could silently fall back from CPU time to wall time, according to os_linux.cpp. We'd rather to have true CPU time, or nothing. Mixing up CPU time and wall time is confusing to users.
- `_vtime_accum` only tracks the time consumed by the concurrent marking main thread, but not the concurrent worker threads.

There's a `G1ConcurrentMarkThread::vtime_accum()` that seems to account for concurrent worker threads. It looks only used for logging. It might be possible to replace the existing logging code with the value of the `sun.threads.cpu_time.gc_conc_mark` hsperf counter. However, it is better to do that separately, and I'll create an RFE.

> I believe the purpose of calling update_concurrent_mark_threads_cpu_time in multiple places is to get more up-to-date conc-cpu-time. Reading through the JBS ticket, I don't see the motivation for maintaining such a "fresh" value.

Yes. The reason is that a concurrent mark cycle could take several minutes for a large heap. For tools like [AHS](https://mail.openjdk.org/pipermail/hotspot-dev/2022-September/064190.html) that reads these CPU hsperf counters, they could read these CPU data every 1 to 5 seconds.

The current mechanism still needs to wait for the whole `G1ConcurrentMark::mark_from_roots()` to complete, which could still take minutes. I can create a separate RFE to make it update more frequently inside `G1CMConcurrentMarkingTask::work()`.

caoman replied:
I've created https://bugs.openjdk.org/browse/JDK-8318937 and https://bugs.openjdk.org/browse/JDK-8318941.

dholmes-ora commented:
> `os::elapsedVTime()` could silently fall back from CPU time to wall time, according to os_linux.cpp

The only allowed failure modes for `getrusage` are:

 EFAULT usage points outside the accessible address space.
 EINVAL who is invalid.

neither of which are going to occur in practice

albertnetymk commented:
> For tools like AHS that reads these CPU hsperf counters, they could read these CPU data every 1 to 5 seconds.

Okay, these counters can be accessed frequently, but is it necessary for them to provide up-to-date information on every access? If not, what level of delay is acceptable? I assume this depends on how often AHS resizes the heap. (In the Parallel case, I believe the counters can be outdated for the duration of the full-gc.)

My primary concern is that the change in G1 is too intrusive -- the logic for tracking/collecting thread-CPU is scattered in many places. Additionally, the situation is expected to worsen in the near future, based on the statement "I can create a separate RFE to make it update more frequently..."

Also, why isn't `G1ServiceThread` part of the change? I would expect all subclasses of `ConcurrentGCThread` to be taken into account. Is this omission intentional?

Finally, thread-CPU time is something tracked at the OS level, so it's a bit odd that one has to instrument the VM to get that information.

> - Looking into the `/proc/` method, it also is not super intuitive as to which value corresponds to GC time, and it might be difficult to reliably identify all PIDs of the relevant GC threads.

According to https://man7.org/linux/man-pages/man5/proc.5.html, "(14) utime" + "(15) stime  %lu" == thread-cpu-time. `cat /proc/<java-pid>/task/*/stat` lists all VM internal threads, including GC, JIT, and etc.

jjoo172 commented:
> Okay, these counters can be accessed frequently, but is it necessary for them to provide up-to-date information on every access? If not, what level of delay is acceptable? I assume this depends on how often AHS resizes the heap. (In the Parallel case, I believe the counters can be outdated for the duration of the full-gc.)

The way AHS is implemented, we access these counters upon every heap resize, so in theory every call to `expand()` and `shrink()` will rely to some extent on these counters. If the counters do happen to be stale, AHS won't break, but will certainly be less effective.

For the rest of the points, Man do you have any additional insight?

caoman commented:
> Okay, these counters can be accessed frequently, but is it necessary for them to provide up-to-date information on every access? If not, what level of delay is acceptable?  (In the Parallel case, I believe the counters can be outdated for the duration of the full-gc.)

Besides Jonathan's point, in our experience updating these counters once every 1 second is good enough for AHS. It might even be OK for once every 2-3 seconds. We just don't want the counters to be outdated for tens or hundreds of seconds. Also the "tens or hundreds of seconds" delay is mainly a problem for concurrent mark, but less of a problem for GC pauses like the full-GC, because:
- Multi-second GC pauses are uncommon compared to multi-second concurrent mark.
- GC pauses are frequent. It is OK for AHS to get slightly outdated info for the ongoing GC pause, because AHS can quickly influence the next GC pause after the ongoing GC pause finishes and updates the counters. This is also the reason we only refresh `sun.threads.total_gc_cpu_time` after a GC pause.

> My primary concern is that the change in G1 is too intrusive -- the logic for tracking/collecting thread-CPU is scattered in many places. Additionally, the situation is expected to worsen in the near future, based on the statement "I can create a separate RFE to make it update more frequently..."

I think most G1 changes in this PR are straightforward and easy to maintain. With the exception of concurrent mark (`sun.threads.cpu_time.gc_conc_mark`), each counter is only updated at exactly one place. As part of [JDK-8318941](https://bugs.openjdk.org/browse/JDK-8318941), we could find a way to update `sun.threads.cpu_time.gc_conc_mark` at only one place as well.

In addition, we think it is well worth the effort for G1 (or any modern garbage collector) to keep track of CPU time spent by their GC threads. Besides monitoring benefits of users and external tools like AHS, it could open up opportunity for G1 to develop better heuristics based on CPU time. We find CPU time spent by GC threads is a better measure for GC overhead, than wall (pause) time. There was [a discussion](https://mail.openjdk.org/pipermail/hotspot-gc-dev/2021-May/035241.html) about  `GCTimeRatio` and CPU time. Today even after [JDK-8253413](https://bugs.openjdk.org/browse/JDK-8253413), `GCTimeRatio` still only accounts for pause time. We hope JVM could provide a flag `GCCpuRatio` and resizes its heap to respect `GCCpuRatio`. AHS actually tries to partly achieve the effect of a `GCCpuRatio` flag.

> Also, why isn't `G1ServiceThread` part of the change? I would expect all subclasses of ConcurrentGCThread to be taken into account. Is this omission intentional?

Thanks. We missed this thread in our accounting due to oversight. It was named `G1YoungRemSetSamplingThread` before. @jjoo172, could you add an hsperf counter for `G1ServiceThread`?

> Finally, thread-CPU time is something tracked at the OS level, so it's a bit odd that one has to instrument the VM to get that information.
> According to https://man7.org/linux/man-pages/man5/proc.5.html, "(14) utime" + "(15) stime %lu" == thread-cpu-time. cat `/proc/<java-pid>/task/*/stat` lists all VM internal threads, including GC, JIT, and etc.

It is possible, but it is a lot of work for the users. `/proc/<java-pid>/task/*/stat` lists all Java threads from the application as well. Users would need deep JVM knowledge to find out which threads are GC threads, which are JIT threads, etc. Quite a few other issues come along as well:
- What if the JVM's internal threads are renamed across different JDK versions? (E.g. `GC Thread#N` was named `Gang worker#N` in JDK 8.)
- Different tools that need to read CPU time would each need to implement a parser for `/proc/<java-pid>/task/*/stat`, and deal with similar problems.
- What if a tool needs to support multiple OSes like Windows, which do not have `/proc` FS?

caoman reviewed (approved):
LGTM still, thanks!

kstefanj commented:
Sorry for being a bit late to this PR. I think the addition of CPU time tracking is good, but I wonder if we could do it in a way that is a bit more general. A more general way of tracking CPU time for a set of threads and we could then have different consumers of this data. In addition to hsperf counters I think having logging and JFR events for this could be interesting as well. 

Have you had any thought along those lines, any obvious problems with such approach?

jjoo172 commented:
@kstefanj thank you for taking a look!  caoman@ and I discussed your comment and our thoughts are below:

> A more general way of tracking CPU time for a set of threads and we could then have different consumers of this data.

Could you elaborate a bit on what you were thinking of here? If we are assuming something like a thread that updates all other threads, I think this implementation could get a bit complicated.

There are two main issues that we can see with a generic thread approach:
1. We would have to figure out how often to pull metrics from the various gc threads from the central thread, and possibly determine this frequency separately for every thread. Instead with our current implementation, we can manually trigger publishes based on when the GC thread is done doing work.
2.  We would still need to tag each thread we want to track somewhere, and keep track of a mapping from thread to its counter name, etc. which doesn't seem to simplify things too much. (I imagine we will still need to touch numerous files to "tag" each thread with whether we want to track it or not?)

The existing `sun.management:type=HotspotThreading` MBean approach (discussed [here]( https://mail.openjdk.org/pipermail/core-libs-dev/2023-September/111397.html)) could be another general way to track CPU. However, the discussion concludes that it is an internal API, and discourages users from using it.

> In addition to hsperf counters I think having logging and JFR events for this could be interesting as well.

There's actually already an existing implementation that covers process CPU from GC pauses, but it doesn't handle concurrent work: https://bugs.openjdk.org/browse/JDK-8291753. However, for our implementation of AHS, we decided against a JFR-based approach, as it has a bit more overhead and the hsperf-based implementation seemed simpler. So while a JFR-based approach to track these counters might be feasible, we believe it would be considerably more work, and could be done in a separate PR later if there is sufficient user interest.

simonis reviewed (approved):
@kstefanj , it is a pity that the `sun.management:type=HotspotThreading` MBean is not exported any more. If we move that or a similar functionality to a new MBean under `com.sun.management` (as proposed in the [cited discussion](https://mail.openjdk.org/pipermail/core-libs-dev/2023-September/111397.html)) then we might reuse these new hsperf counters in the same way this is already done by some other MBeans which already use hsperf counters as their information source. I think logging or JFR functionality could also easily be implemented on top of the new hsperf counters.

As for @albertnetymk's proposal to simple use the `/proc` file system to retrieve thread CPU time information, that is not only inconvenient for the reasons detailed by @caoman. It will also not work for compiler threads (which I think should have similar counters in the future) because compiler threads can be created and and destroyed dynamically (due to [`-XX:+UseDynamicNumberOfCompilerThreads`](https://bugs.openjdk.org/browse/JDK-8198756)).

The current PR still looks good to me.

kstefanj commented:
> The existing sun.management:type=HotspotThreading MBean approach (discussed [here](https://mail.openjdk.org/pipermail/core-libs-dev/2023-September/111397.html)) could be another general way to track CPU. However, the discussion concludes that it is an internal API, and discourages users from using it.

> @kstefanj , it is a pity that the `sun.management:type=HotspotThreading` MBean is not exported any more. If we move that or a similar functionality to a new MBean under `com.sun.management` (as proposed in the [cited discussion](https://mail.openjdk.org/pipermail/core-libs-dev/2023-September/111397.html)) then we might reuse these new hsperf counters in the same way this is already done by some other MBeans which already use hsperf counters as their information source. I think logging or JFR functionality could also easily be implemented on top of the new hsperf counters.

I haven't looked at the details around this, but extracting some useful information from the internal bean and making it public sound reasonable to me.

caoman commented:
> I played around a bit instead of trying to explain what I mean and this is not very polished, but I was thinking something like this: https://github.com/openjdk/jdk/compare/pr/15082...kstefanj:jdk:pull/15082-idea
>
> What do you think? This way we don't add things to CollectedHeap as well, which is usually good unless really needed.

I think it looks great. It is mainly refactoring that consolidates the declarations/definitions of the hsperf counters in to a single file. Would it be better to name that class `CPUTimeCounters`, so we could move `sun.threads.cpu_time.vm` and `sun.threads.cpu_time.conc_dedup`, and future JIT thread CPU counters to that class?

Then we could also change the constructor of `ThreadTotalCPUTimeClosure` to `ThreadTotalCPUTimeClosure(CPUTimeCounters* counters, CPUTimeGroups::Name name)`, then it could set `_update_gc_counters` based on `name`.

kstefanj commented:
> I think it looks great. It is mainly refactoring that consolidates the declarations/definitions of the hsperf counters in to a single file. Would it be better to name that class `CPUTimeCounters`, so we could move `sun.threads.cpu_time.vm` and `sun.threads.cpu_time.conc_dedup`, and future JIT thread CPU counters to that class?
>
Yes, mainly refactoring and I was thinking along the same lines, but since this patch was just for GC and we had `CollectorCounters` already I went with this. I think calling the class `CPUTimeCounters` would be good and place it outside GC makes sense if we plan to include even more CPU time counters.

Another name that we could improve is `CPUTimeGroups` and maybe also the enum name `Name`, they are ok, but we might come up with something better. 

 
> Then we could also change the constructor of `ThreadTotalCPUTimeClosure` to `ThreadTotalCPUTimeClosure(CPUTimeCounters* counters, CPUTimeGroups::Name name)`, then it could set `_update_gc_counters` based on `name`.

I was looking at this too, but had to restructure the code more to avoid circular deps. If we create the general `CPUTImeCounters` we could move this closure to that file and then things would fit better I believe. So I like your proposals.

jjoo172 commented:
/contributor add @kstefanj

kstefanj commented on src/hotspot/share/gc/g1/g1ConcurrentRefineThread.cpp:
```diff
@@ -79,6 +79,7 @@ void G1ConcurrentRefineThread::run_service() {
     } else {
       _vtime_accum = 0.0;
     }
+    maybe_update_threads_cpu_time();
```
I think the lines above here: 
```
if (os::supports_vtime()) {
  _vtime_accum = (os::elapsedVTime() - _vtime_start);
} else {
  _vtime_accum = 0.0;
}
```
Should be extracted out into the new method and instead of calling it `maybe_update_threads_cpu_time()` just call `track_usage()` or `track_cpu_time()`. The the implementation in the primary thread can then call this and do the extra tracking.

kstefanj commented on src/hotspot/share/gc/g1/g1ServiceThread.cpp:
```diff
@@ -130,6 +131,12 @@ void G1ServiceThread::run_task(G1ServiceTask* task) {
 
   task->execute();
 
+  if (UsePerfData && os::is_thread_cpu_time_supported()) {
+    CPUTimeCounters* counters = G1CollectedHeap::heap()->cpu_time_counters();
+    ThreadTotalCPUTimeClosure tttc(counters, CPUTimeGroups::gc_service);
+    tttc.do_thread(task->_service_thread);
+  }
```
Please extract this to a function, similar to the other cases something like `track_cpu_time()`.

kstefanj reviewed (changes requested):
A few more comments.

kstefanj commented on src/hotspot/share/gc/g1/g1ConcurrentMark.cpp:
```diff
@@ -2080,6 +2080,18 @@ void G1ConcurrentMark::print_on_error(outputStream* st) const {
   _mark_bitmap.print_on_error(st, " Bits: ");
 }
 
+void G1ConcurrentMark::update_concurrent_mark_threads_cpu_time() {
+  assert(Thread::current() == cm_thread(),
+         "Must be called from _cmThread to avoid races");
+  if (!UsePerfData || !os::is_thread_cpu_time_supported()) {
+    return;
+  }
+  CPUTimeCounters* counters = _g1h->cpu_time_counters();
+  ThreadTotalCPUTimeClosure tttc(counters, CPUTimeGroups::gc_conc_mark);
+  tttc.do_thread(cm_thread());
+  threads_do(&tttc);
+}
```
Any particular reason for having this in `G1ConcurrentMark` instead of keeping it where it is called in `G1ConcurrentMarkThread`. The implementation would be more or less the same apart from the two last lines:
```
  tttc.do_thread(this);
  cm()->threads_do(&tttc);
```

kstefanj commented:
One more thing I noticed is that the parallel worker time from the Remark pause is not updated after the remark pause. So I guess we should add a call to update that. It will still be account the next time we do a normal GC, but if we want the counter to be up to date as much as possible we should update it after the remark pause.

jjoo172 commented on src/hotspot/share/gc/g1/g1ConcurrentRefineThread.cpp:
```diff
@@ -182,6 +184,12 @@ void G1PrimaryConcurrentRefineThread::do_refinement_step() {
 
+void G1PrimaryConcurrentRefineThread::maybe_update_threads_cpu_time() {
+  if (UsePerfData && os::is_thread_cpu_time_supported()) {
+    cr()->update_concurrent_refine_threads_cpu_time();
```
Thank you for the review! Do you think you could take a look at the newest update and see if that aligns with what you were thinking? I wanted to make sure I understood your comments correctly.

jjoo172 commented:
Will address the Remark pause update in my next commit!

jjoo172 commented:
I believe all comments have been addressed and this PR is once again RFR!

kstefanj commented:
Thanks for addressing my comments. I have a few more things:

- I think all changes to `test_g1ServiceThread.cpp` can be reverted. Should not be needed now
- Please fix all whitespace issues
- Should we move the VMThread and StringDedup counters into `CPUTimeCounters` as well? Any problem with this?

caoman commented on src/hotspot/share/runtime/init.cpp:
```diff
@@ -120,6 +121,8 @@ jint init_globals() {
   compilationPolicy_init();
   codeCache_init();
   VM_Version_init();              // depends on codeCache_init for emitting code
+  // Initialize CPUTimeCounters object, which must be done before creation of the heap.
```
Would it be possible to move this inside `universe_init()` in universe.cpp, somewhere above `Universe::initialize_heap()`? There's a similar `MetaspaceCounters::initialize_performance_counters()` in `universe_init()`.

caoman commented on src/hotspot/share/gc/g1/g1CollectedHeap.hpp:
```diff
@@ -56,6 +56,7 @@
 #include "memory/allocation.hpp"
 #include "memory/iterator.hpp"
 #include "memory/memRegion.hpp"
+#include "runtime/cpuTimeCounters.hpp"
```
Probably move this include to the .cpp file?

jjoo172 commented:
@kstefanj Good points - done!

caoman commented on src/hotspot/share/gc/g1/g1ServiceThread.cpp:
```diff
@@ -163,6 +155,13 @@ void G1ServiceThread::stop_service() {
+void G1ServiceThread::update_thread_cpu_time(G1ServiceTask* task) {
+  if (UsePerfData && os::is_thread_cpu_time_supported()) {
+    ThreadTotalCPUTimeClosure tttc(CPUTimeCounters::get_instance(), CPUTimeGroups::gc_service);
+    tttc.do_thread(task->_service_thread);
```
It could just use `do_thread(this)`, then it can remove the `task` parameter.

caoman commented on src/hotspot/share/gc/shared/stringdedup/stringDedupProcessor.cpp:
```diff
@@ -69,9 +70,9 @@ void StringDedup::Processor::initialize() {
   _processor = new Processor();
   if (UsePerfData && os::is_thread_cpu_time_supported()) {
     EXCEPTION_MARK;
```
This whole `if` block could be updated to `CPUTimeCounters::get_instance()->create_counter(CPUTimeGroups::conc_dedup)`. We can also remove the `_concurrent_dedup_thread_cpu_time` field and the `ThreadTotalCPUTimeClosure(PerfCounter* counter)` constructor.

In `StringDedup::Processor::run()`, it can call
```
if (UsePerfData && os::is_thread_cpu_time_supported()) {
  ThreadTotalCPUTimeClosure tttc(CPUTimeCounters::get_instance(), CPUTimeGroups::conc_dedup);
  tttc.do_thread(thread);
}
```

Similar, this can be applied to vmThread.

caoman reviewed (changes requested):
Overall looks good, a few details could be improved.

jjoo172 commented:
Addressed comments, but am running into an assertion failure here when building: https://github.com/openjdk/jdk/pull/15082/files#diff-d1c5f7a125171a3828bdb3f4488327e03ceb96ff615eacfc9df406aca408cbe6R72

Will investigate tomorrow.

jjoo172 commented:
All comments have been addressed, this PR should once again be RFR.

caoman commented on src/hotspot/share/gc/g1/g1ConcurrentMark.cpp:
```diff
@@ -68,6 +68,7 @@
 #include "oops/access.inline.hpp"
 #include "oops/oop.inline.hpp"
 #include "runtime/atomic.hpp"
+#include "runtime/cpuTimeCounters.hpp"
```
This include could be removed.

caoman commented on src/hotspot/share/gc/g1/g1ConcurrentRefine.cpp:
```diff
@@ -35,6 +35,7 @@
 #include "logging/log.hpp"
 #include "memory/allocation.inline.hpp"
 #include "memory/iterator.hpp"
+#include "runtime/cpuTimeCounters.hpp"
```
This could be removed too.

caoman commented on src/hotspot/share/gc/g1/g1ServiceThread.cpp:
```diff
@@ -23,8 +23,10 @@
  */
 
 #include "precompiled.hpp"
+#include "gc/g1/g1CollectedHeap.hpp"
```
Could this include be removed?

caoman commented on src/hotspot/share/gc/g1/g1ServiceThread.hpp:
```diff
@@ -27,6 +27,7 @@
 
 #include "gc/shared/concurrentGCThread.hpp"
 #include "runtime/mutex.hpp"
+#include "runtime/perfData.hpp"
```
This could be removed.

caoman commented on src/hotspot/share/gc/shared/collectedHeap.cpp:
```diff
@@ -49,6 +49,7 @@
 #include "memory/universe.hpp"
 #include "oops/instanceMirrorKlass.hpp"
 #include "oops/oop.inline.hpp"
+#include "runtime/atomic.hpp"
```
This could be removed.

caoman commented on src/hotspot/share/gc/shared/stringdedup/stringDedupProcessor.cpp:
```diff
@@ -25,6 +25,7 @@
 #include "precompiled.hpp"
 #include "classfile/javaClasses.inline.hpp"
 #include "classfile/stringTable.hpp"
+#include "gc/shared/collectedHeap.hpp"
```
This include as well as the include for perfData.hpp (line 45) could be removed.

caoman commented on src/hotspot/share/gc/shared/stringdedup/stringDedupProcessor.cpp:
```diff
@@ -64,6 +67,10 @@ StringDedup::Processor::Processor() : _thread(nullptr) {}
 
 void StringDedup::Processor::initialize() {
   _processor = new Processor();
+  if (UsePerfData && os::is_thread_cpu_time_supported()) {
```
The if and EXCEPTION_MARK could be removed, because `create_counter()` does that internally.

caoman commented on src/hotspot/share/runtime/vmThread.cpp:
```diff
@@ -136,6 +137,9 @@ void VMThread::create() {
     _perf_accumulated_vm_operation_time =
                  PerfDataManager::create_counter(SUN_THREADS, "vmOperationTime",
                                                  PerfData::U_Ticks, CHECK);
+    if (os::is_thread_cpu_time_supported()) {
```
The if could be remove, as `create_counter()` checks it internally.

caoman commented on src/hotspot/share/gc/shared/stringdedup/stringDedupProcessor.hpp:
```diff
@@ -27,6 +27,7 @@
 
 #include "gc/shared/stringdedup/stringDedup.hpp"
 #include "memory/allocation.hpp"
+#include "runtime/perfData.hpp"
```
This could be removed.

caoman reviewed (approved):
Looks pretty clean. Only a few minor cleanups remain.

jjoo172 commented:
Addressed cleanups, PR should be RFR!

kstefanj commented on src/hotspot/share/runtime/cpuTimeCounters.hpp:
```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (c) 2002, 2023, Oracle and/or its affiliates. All rights reserved.
```
This is probably on me, but this 2002 is wrong:
```suggestion
 * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
```

kstefanj reviewed (commented):
I think this looks good now. I've started some additional testing to make sure we don't run into anything unexpected with the newly added test on any platforms. 

Will approve once the testing is green.

albertnetymk commented on src/hotspot/share/runtime/cpuTimeCounters.cpp:
```diff
@@ -0,0 +1,130 @@
+}
+
+ThreadTotalCPUTimeClosure::~ThreadTotalCPUTimeClosure() {
+    CPUTimeCounters* instance = CPUTimeCounters::get_instance();
+    PerfCounter* counter = instance->get_counter(_name);
+    jlong net_cpu_time = _total - counter->get_value();
+    counter->inc(net_cpu_time);
+    if (CPUTimeGroups::is_gc_counter(_name)) {
+      instance->inc_gc_total_cpu_time(net_cpu_time);
+    }
```
I feel much of this is on the wrong abstraction level; `CPUTimeCounters::update_counter(_name, _total);` should be sufficient here. (The caller handles diff calculation and inc gc-counter if needed.)

kstefanj replied:
We could add a new closure just used by GC that 's a sub-class of `ThreadTotalCPUTimeClosure` and just adds this to the constructor: 
```
instance->inc_gc_total_cpu_time(net_cpu_time);
```

That way we could get rid of `CPUTimeGroups::is_gc_counter()` as well since all those counters should use the "GC closure" or we can keep it and assert that no GC closure uses the wrong closure.

What do you think about that Albert, would that address your concerns?

albertnetymk replied:
(I just realized that I made a typo in my previous msg; should be *callee* instead.) That is what I have in mind.

```cpp
void CPUTimeCounters::update_counter(name, total) {
  auto counter = get_counter(name);
  auto old_v = counter->get_value();
  auto diff = total - old_v;
  counter->inc(diff);
  if (counter->is_gc_counter()) {
    counter->inc(diff);
  }
}
```

jjoo172 replied:
I'm not sure I understood correctly, could you let me know if this latest commit addresses your comment in the way you were intending?

kstefanj replied:
It does.

albertnetymk commented on src/hotspot/share/runtime/cpuTimeCounters.hpp:
```diff
@@ -0,0 +1,109 @@
+  void create_counter(CounterNS ns, CPUTimeGroups::CPUTimeType name);
+
+public:
+  static CPUTimeCounters* get_instance() {
+    assert(_instance != nullptr, "no instance found");
+    return _instance;
+  }
```
Seems that this is needed solely for accessing the following instance methods. I wonder if it's possible to expose only static methods in the public API.

kstefanj commented on src/hotspot/share/runtime/cpuTimeCounters.cpp:
```diff
@@ -0,0 +1,130 @@
+  Atomic::add(&_gc_total_cpu_time_diff, diff);
+}
+
+void CPUTimeCounters::publish_gc_total_cpu_time() {
+  // Ensure that we are only incrementing atomically by using Atomic::cmpxchg
+  // to set the value to zero after we obtain the new CPU time difference.
+  jlong old_value;
+  jlong fetched_value = Atomic::load(&_gc_total_cpu_time_diff);
+  jlong new_value = 0;
+  do {
+    old_value = fetched_value;
+    fetched_value = Atomic::cmpxchg(&_gc_total_cpu_time_diff, old_value, new_value);
+  } while (old_value != fetched_value);
+  get_counter(CPUTimeGroups::CPUTimeType::gc_total)->inc(fetched_value);
+}
```
Why do we have to do this publish dance? Couldn't the closure that update the diff instead just update the counter? From what I can tell we never have multiple closures active at the same time so should be no race there?

albertnetymk replied:
This two-step update does seem unnecessary, IMO.

jjoo172 replied:
I agree that in the case that multiple closures are not active at the same time, we wouldn't have to implement it in this way. However, isn't it possible to have multiple closures active simultaneously, e.g.  vm thread, concurrent mark thread, concurrent refine thread? I think to account for the races there, we can only update the `_gc_total_cpu_time_diff` member variable atomically during these closure destructions, and then publish the actual updated `gc_total` counter at manually specified times via `publish_gc_total_cpu_time()`. If we were to call `publish_gc_total_cpu_time` as part of the thread closure, I believe it would be difficult to prevent races when accessing the underlying counter from the various gc-related threads.

Or maybe there is another strategy that I'm not seeing?

albertnetymk replied:
Both `publish_gc_total_cpu_time` and `~ThreadTotalCPUTimeClosure` are called by the vm-thread inside a safepoint, so there shouldn't be any other threads running simultaneously, I believe.

kstefanj replied:
Me and Albert just spoke and we do see the problem that two concurrent threads could be executing the closure at the same time. So if having a total counter we need to sync the updates. But when talking we started to question how useful it is to have the gc_total counter. It is just an aggregate of the other gc-counters, but it is out of sync between safepoints. So you will always get a more accurate value by looking at the individual gc-counters.

We came to the conclusion that it would probably be easier to drop `gc_total` right now, rather than trying to keep it in sync for all updates to the individual counters. Because having them out of sync doesn't feel like a great option. 

Are we missing anything or do you agree?

jjoo172 replied:
@simonis was the original suggester of this counter, so I will defer to his expertise. I do agree that dropping the counter would simplify things, but it also might not hurt to just leave it in. I'm okay with either option!

caoman replied:
I think the ideal approach to simplify this is to support Atomic operation on a `PerfCounter`.
We could either introduce a `PerfAtomicCounter`/`PerfAtomicLongCounter` class, or perform `Atomic::add()` on the `PerfData::_valuep` pointer. There's already `PerfData::get_address()`, so we might be able to do:

```
Atomic::add((volatile jlong *)(instance->get_counter(CPUTimeGroups::CPUTimeType::gc_total)->get_address()), net_cpu_time);
```

However, a new class `PerfAtomicCounter`  is likely cleaner. E.g., we may also want to make `PerfAtomicCounter::sample()` use a CAS. It is probably better to introduce `PerfAtomicCounter` in a separate RFE later.

Would the `Atomic::add()` with `PerfData::get_address()` approach be OK for now, or would we rather introduce a lock, or leave the `gc_total` mechanism as-is and address the out-of-sync-ness in a follow-up RFE?

IMO the out-of-sync-ness problem is minor, because users are likely to either look at a single `gc_total` counter, or look at each individual GC  CPU counter and disregard `gc_total`.

jjoo172 replied:
In the interest of the RDP1 deadline, should we leave improving the sync issues with gc_total to a separate RFE? (Especially given that a "correct" design may take some time to come up with, and that gc_total being slightly out of sync is not a major issue.)

caoman replied:
Thanks, opened https://bugs.openjdk.org/browse/JDK-8321304.

kstefanj reviewed (commented):
From a testing point of view I think this is looking good now, re-ran the failing test and some other jstat tests as well and they all pass.

Please address the comments from Albert and we can hopefully finish this before RDP1.

jjoo172 commented on src/hotspot/share/gc/g1/g1CollectedHeap.cpp:
```diff
@@ -2414,6 +2423,24 @@ void G1CollectedHeap::verify_region_attr_remset_is_tracked() {
+    return;
+  }
+  WorkerThreads* worker_threads = workers();
+  if (worker_threads != nullptr) {
```
I guess it shouldn't be, since it is set when we call initialize() and also never unset. I'll remove this conditional.

kstefanj commented on src/hotspot/share/gc/parallel/parallelScavengeHeap.cpp:
```diff
@@ -884,3 +889,18 @@ void ParallelScavengeHeap::pin_object(JavaThread* thread, oop obj) {
+  // behavior, we should rethink if it is still safe.
+  gc_threads_do(&tttc);
+
+  CPUTimeCounters::publish_gc_total_cpu_time();
```
As I suggested in the other comment, maybe we should not keep the total counter, but if we do we need to make sure the destructor of the closure is run before the call to `publish_gc_total_cpu_time()`. Otherwise we will publish a not yet updated value.

simonis replied:
I still think that a total counter is useful and I'd appreciate if you can keep it. To second what @caoman said before, it is GC agnostic, easy to use even for non GC experts and future proof with regards to implementation changes in the GCs. Please keep it.

jjoo172 replied:
Put the closure in a scope, I think that should address the concern.

kstefanj replied:
Yes, scoping it will work.

kstefanj reviewed (changes requested):
Few more comments after the latest changes.

caoman commented on src/hotspot/share/runtime/cpuTimeCounters.hpp:
```diff
@@ -0,0 +1,112 @@
+  static bool is_gc_counter(CPUTimeType val);
+};
+
+class CPUTimeCounters: public CHeapObj<mtServiceability> {
+private:
+  // CPUTimeCounters is a singleton instance.
+  CPUTimeCounters();
+  NONCOPYABLE(CPUTimeCounters);
+
+  static CPUTimeCounters* _instance;
```
I thought it is typically preferred to initialize a singleton object on the heap, rather than using many static variables. It is easier to control the initialization order and timing of an on-heap singleton object than statics.

Moreover, for this class, `initialize()` could also check `if (UsePerfData)`, and only create the singleton object under `UsePerfData`. This could save some memory when `UsePerfData` is false.

kstefanj replied:
I would say it depends on the use-case and here when switching to use static functions to use the instance it felt more like an all-static class. I agree that it would be nice to avoid the additional memory usage if `UsePerfData` is `false` so I'm ok with keeping the instance if we add that.

jjoo172 replied:
Added the check in the initializer

jjoo172 commented:
/integrate

caoman commented:
/sponsor
